### ISSUE TITLE
Module 4: Evaluation Metrics - IoU & Dice (1 Pomodoro)

### ISSUE THREAD

**Issue by gregory1506** [gregory1506/arc-pytorch-prep#4]
## Module 4: Evaluation Metrics - IoU & Dice
**Duration**: 1 Pomodoro (25 minutes)  
**Status**: 🔄 Not Started  
**Priority**: Medium
**Depends on**: #2

### Learning Objectives
By the end of this module, students will be able to:
1. Calculate Intersection over Union (IoU / Jaccard Index)
2. Calculate Dice coefficient (F1 score for segmentation)
3. Implement per-class and mean metrics
4. Handle edge cases (empty masks, single class)
5. Choose appropriate metric for their problem

### Full Pomodoro: Teaching + Implementation + Testing
- [ ] **Teaching (8 min)**: Metrics for semantic segmentation
- [ ] **Code (10 min)**: Fill in src/evaluation/metrics.py
  - TODO 4.1: Implement IoU calculation
  - TODO 4.2: Implement Dice coefficient
  - TODO 4.3: Add per-class and mean metrics
  - TODO 4.4: Handle edge cases (empty masks)
- [ ] **Test (5 min)**: Run pytest tests/test_module_04_metrics.py -v
- [ ] **Verification (2 min)**: Test with known values

### Files to Edit
- src/evaluation/metrics.py

### Test Command
pytest tests/test_module_04_metrics.py -v

### Teaching Notes
See TEACHING.md#module-4-evaluation-metrics for:
- IoU vs Dice comparison
- Multi-class metric strategies
- Edge case handling

### Acceptance Criteria
- [ ] IoU = 1.0 for perfect overlap
- [ ] IoU = 0.0 for no overlap
- [ ] Dice = 2*IoU/(1+IoU) relationship holds
- [ ] Empty masks handled with smoothing
- [ ] Multi-class metrics work correctly
- [ ] All tests in test_module_04_metrics.py pass

### Resources
- Metrics for Semantic Segmentation: https://www.jeremyjordan.me/evaluating-image-segmentation-models/
- Mean IoU Explained: https://www.tensorflow.org/api_docs/python/tf/keras/metrics/MeanIoU

---
**Labels**: tutorial, module-4, teaching, coding, testing
